python/ray/data/_internal/arrow_ops/transform_pyarrow.py

```diff
@@ -267,6 +267,19 @@ def unify_schemas(
     else:
         schemas_to_unify = schemas
 
+    # Promote nested struct fields containing tensors to variable-shaped if needed
+    for s_idx, schema in enumerate(schemas_to_unify):
```

Comment:
Let's try to restructure this method such that

 - Unifying is happening per column
 - We have a method that unifies all schemas for a given column
 - We use that method recursively for struct-types as well